app/controllers/PdfDownloadController.scala

```diff
@@ -39,47 +40,49 @@ class PdfDownloadController @Inject()(appConfig: FrontendAppConfig,
                                      ) extends FrontendController with I18nSupport {
 
   def application(reference: String, token: Option[String]): Action[AnyContent] = identify.async { implicit request =>
-    if(request.eoriNumber.isDefined) {
-      getApplicationPDF(request.eoriNumber.get, reference)
-    } else if(token.isDefined) {
-      pdfService.decodeToken(token.get) match {
-        case Some(eori) => getApplicationPDF(eori, reference)
-        case None => Future.successful(Redirect(controllers.routes.SessionExpiredController.onPageLoad()))
-      }
-    } else {
-      Future.successful(Redirect(controllers.routes.UnauthorisedController.onPageLoad()))
-    }
+    getPdf(reference, token, getApplicationPDF)
   }
 
   def ruling(reference: String, token: Option[String]): Action[AnyContent] = identify.async { implicit request =>
-    if(request.eoriNumber.isDefined) {
-      getRulingPDF(request.eoriNumber.get, reference)
-    } else if(token.isDefined) {
-      pdfService.decodeToken(token.get) match {
-        case Some(eori) => getRulingPDF(eori, reference)
+    getPdf(reference, token, getRulingPDF)
+  }
+
+  private def getPdf(reference: String, token: Option[String], pdf: (String, String) => Future[Result])
+                    (implicit request: IdentifierRequest[AnyContent]): Future[Result] = {
+    (request.eoriNumber, token) match {
+      case (Some(eori), _) => pdf(eori, reference)
+      case (_, Some(tkn)) => pdfService.decodeToken(tkn) match {
+        case Some(eori) => pdf(eori, reference)
         case None => Future.successful(Redirect(controllers.routes.SessionExpiredController.onPageLoad()))
       }
-    } else {
-      Future.successful(Redirect(controllers.routes.UnauthorisedController.onPageLoad()))
+      case _ => Future.successful(Redirect(controllers.routes.UnauthorisedController.onPageLoad()))
+    }
+  }
+
+  private def getRulingPDF(implicit request: Request[AnyContent]): (String, String) => Future[Result] = { (eori, reference) =>
+    caseService.getCaseWithRulingForUser(eori, reference) flatMap { c: Case =>
+      generatePdf(
+        rulingPdf(appConfig, c, c.decision.getOrElse(throw new IllegalStateException("Missing decision"))), s"BTIRuling$reference.pdf"
+      )
     }
   }
 
-  private def getApplicationPDF(eori: String, reference: String)(implicit r: Request[_]): Future[Result] = {
+  private def getApplicationPDF(implicit request: Request[AnyContent]): (String, String) => Future[Result] = { (eori, reference) =>
```

Comment:
I think the signature of this was better before. Its not very intuitive now

Reply:
I know, but I need a function if I want to remove repetitions from the code

Reply:
what do u suggest?

Reply:
It is a function, you can just change the way it is called.

Reply:
How do you like it?